deepchem/models/normalizing_flows.py

```diff
@@ -0,0 +1,297 @@
+"""
+Normalizing flows for transforming probability distributions.
+"""
+
+import numpy as np
+import logging
+from typing import List, Iterable, Optional, Tuple, Sequence, Any
+
+import tensorflow as tf
+from tensorflow.keras.layers import Lambda
+
+import deepchem as dc
+from deepchem.models.losses import Loss
+from deepchem.models.models import Model
+from deepchem.models.keras_model import KerasModel
+from deepchem.models.optimizers import Optimizer, Adam
+
+logger = logging.getLogger(__name__)
+
+
+class NormalizingFlow(tf.keras.models.Model):
+  """Base class for normalizing flow.
+
+  The purpose of a normalizing flow is to map a simple distribution (that is
+  easy to sample from and evaluate probability densities for) to a more
+  complex distribution that is learned from data. The base distribution 
+  p(x) is transformed by the associated normalizing flow y=g(x) to model the
+  distribution p(y).
+
+  Normalizing flows combine the advantages of autoregressive models
+  (which provide likelihood estimation but do not learn features) and
+  variational autoencoders (which learn feature representations but
+  do not provide marginal likelihoods).
+
+  """
+
+  def __init__(self, base_distribution, flow_layers, **kwargs):
+    """Create a new NormalizingFlow.
+
+    Parameters
+    ----------
+    base_distribution : tfd.Distribution
+      Probability distribution to be transformed.
+      Typically an N dimensional multivariate Gaussian.
+    flow_layers : Sequence[tfb.Bijector]
+      An iterable of bijectors that comprise the flow.
+
+    """
+
+    try:
+      import tensorflow_probability as tfp
+      tfd = tfp.distributions
+      tfb = tfp.bijectors
+    except ModuleNotFoundError:
+      raise ValueError(
+          "This class requires tensorflow-probability to be installed.")
+
+    self.base_distribution = base_distribution
+    self.flow_layers = flow_layers
+
+    # Chain of flows is also a normalizing flow
+    bijector = tfb.Chain(list(reversed(self.flow_layers)))
+
+    # An instance of tfd.TransformedDistribution
+    self.flow = tfd.TransformedDistribution(
+        distribution=self.base_distribution, bijector=bijector)
+
+    super(NormalizingFlow, self).__init__(**kwargs)
+
+  def __call__(self, *inputs, training=True):
+    return self.flow.bijector.forward(*inputs)
+
+
+class NormalizingFlowModel(KerasModel):
+  """A base distribution and normalizing flow for applying transformations.
+
+  A distribution implements two main operations:
+    1. Sampling from the transformed distribution.
+    2. Calculating log probabilities.
```

Comment:
Before you dive into the technical details, how about having a paragraph right at the top that describes what sorts of problems normalizing flows are useful for?  Imagine a confused user reading through the documentation.  Help them to figure out right away whether they should keep reading, or go on to some other class that's more likely to be useful for them.